File: java/src/net/razorvine/serpent/DebugVisitor.java
/**
 * Serpent, a Python literal expression serializer/deserializer
 * (a.k.a. Python's ast.literal_eval in Java)
 * Software license: "MIT software license". See http://opensource.org/licenses/MIT
 * @author Irmen de Jong (irmen@razorvine.net)
 */

package net.razorvine.serpent;

import net.razorvine.serpent.ast.*;


/**
 * Ast nodevisitor that prints out the Ast as a string for debugging purposes
 */
public class DebugVisitor implements INodeVisitor
{
	private StringBuilder result = new StringBuilder();
	private int indent=0;
	
	public DebugVisitor()
	{
	}
	
	/**
	 * Get the debug string representation result.
	 */
	@Override
	public String toString()
	{
		return result.toString();
	}
	
	protected void Indent()
	{
		for(int i=0; i<indent; ++i)
			result.append("    ");
	}
	
	public void visit(ComplexNumberNode complex)
	{
		result.append(String.format("complexnumber {0}r {1}i\n", complex.realpart, complex.imaginary));
	}
	
	public void visit(DictNode dict)
	{
		result.append("(dict\n");
		indent++;
		for(INode e: dict.elements)
		{
			Indent();
			KeyValueNode kv = (KeyValueNode) e;
			kv.key.accept(this);
			result.append(" = ");
			kv.value.accept(this);
			result.append(",\n");
		}
		indent--;
		Indent();
		result.append(")");
	}
	
	public void visit(ListNode list)
	{
		result.append("(list\n");
		indent++;
		for(INode node: list.elements)
		{
			Indent();
			node.accept(this);
		}
		indent--;
		Indent();
		result.append(")");
	}
	
	public void visit(NoneNode none)
	{
		result.append("None");
	}
	
	public void visit(IntegerNode value)
	{
		result.append(String.format("int {0}", value.value));
	}
	
	public void visit(LongNode value)
	{
		result.append(String.format("long {0}", value.value));
	}
	
	public void visit(DoubleNode value)
	{
		result.append(String.format("double {0}", value.value));
	}
	
	public void visit(BooleanNode value)
	{
		result.append(String.format("bool {0}", value.value));
	}
	
	public void visit(StringNode value)
	{
		result.append(String.format("string '{0}'", value.value));
	}
	
	public void visit(SetNode setnode)
	{
		result.append("(set\n");
		indent++;
		for(INode node: setnode.elements)
		{
			Indent();
			node.accept(this);
			result.append(",\n");
		}
		indent--;
		Indent();
		result.append(")");
	}
	
	public void visit(TupleNode tuple)
	{
		result.append("(tuple\n");
		indent++;
		for(INode node: tuple.elements)
		{
			Indent();
			node.accept(this);
			result.append(",\n");
		}
		indent--;
		Indent();
		result.append(")");
	}
}

File: dotnet/Serpent/DebugVisitor.cs
/// <summary>
/// Serpent, a Python literal expression serializer/deserializer
/// (a.k.a. Python's ast.literal_eval in .NET)
///
/// Copyright 2014, Irmen de Jong (irmen@razorvine.net)
/// Software license: "MIT software license". See http://opensource.org/licenses/MIT
/// </summary>

using System;
using System.Collections.Generic;
using System.Text;

namespace Razorvine.Serpent
{
	/// <summary>
	/// Ast nodevisitor that prints out the Ast as a string for debugging purposes
	/// </summary>
	public class DebugVisitor: Ast.INodeVisitor
	{
		private StringBuilder result = new StringBuilder();
		private int indent=0;
		
		public DebugVisitor()
		{
		}
		
		/// <summary>
		/// Get the debug string representation result.
		/// </summary>
		public override string ToString()
		{
			return result.ToString();
		}
		
		protected void Indent()
		{
			for(int i=0; i<indent; ++i)
				result.Append("    ");
		}
		
		public void Visit(Ast.ComplexNumberNode complex)
		{
			result.AppendFormat("complexnumber ({0}r,{1}i)", complex.Real, complex.Imaginary);
		}
		
		public void Visit(Ast.DictNode dict)
		{
			result.AppendLine("(dict");
			indent++;
			foreach(Ast.KeyValueNode kv in dict.Elements)
			{
				Indent();
				kv.Key.Accept(this);
				result.Append(" = ");
				kv.Value.Accept(this);
				result.AppendLine(",");
			}
			indent--;
			Indent();
			result.Append(")");
		}
		
		public void Visit(Ast.ListNode list)
		{
			result.AppendLine("(list");
			indent++;
			foreach(Ast.INode node in list.Elements)
			{
				Indent();
				node.Accept(this);
				result.AppendLine(",");
			}
			indent--;
			Indent();
			result.Append(")");
		}
		
		public void Visit(Ast.NoneNode none)
		{
			result.Append("None");
		}
		
		public void Visit(Ast.IntegerNode value)
		{
			result.AppendFormat("int {0}", value.Value);
		}
		
		public void Visit(Ast.LongNode value)
		{
			result.AppendFormat("long {0}", value.Value);
		}
		
		public void Visit(Ast.DoubleNode value)
		{
			result.AppendFormat("double {0}", value.Value);
		}
		
		public void Visit(Ast.BooleanNode value)
		{
			result.AppendFormat("bool {0}", value.Value);
		}
		
		public void Visit(Ast.StringNode value)
		{
			result.AppendFormat("string '{0}'", value.Value);
		}
		
		public void Visit(Ast.DecimalNode value)
		{
			result.AppendFormat("decimal {0}", value.Value);
		}
		
		public void Visit(Ast.SetNode setnode)
		{
			result.AppendLine("(set");
			indent++;
			foreach(Ast.INode node in setnode.Elements)
			{
				Indent();
				node.Accept(this);
				result.AppendLine(",");
			}
			indent--;
			Indent();
			result.Append(")");
		}
		
		public void Visit(Ast.TupleNode tuple)
		{
			result.AppendLine("(tuple");
			indent++;
			foreach(Ast.INode node in tuple.Elements)
			{
				Indent();
				node.Accept(this);
				result.AppendLine(",");
			}
			indent--;
			Indent();
			result.Append(")");
		}
	}
}
